File: 2015/c/day01/main.c
/**
 * main.c
 * 
 * Advent of Code 2015 - Day 1
 * Created By: Samuel Stidham
 *
 * Compiled with: clang -std=c23 -Wall -Wextra -Wpedantic -Werror -O1 -g -fsanitize=address,undefined -fno-omit-frame-pointer main.c -o main
 *                gcc -std=c23 -Wall -Wextra -Wpedantic -Werror -O1 -g -fsanitize=address,undefined -fno-omit-frame-pointer main.c -o main
 */
#define _POSIX_C_SOURCE 200809L

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

void printUsage(char *programName) {
    fprintf(stderr, "Usage: %s -f <input_file>\n", programName);
    exit(EXIT_FAILURE);
}

char* processArguments(int argc, char *argv[]) {
    int opt;

    char *input_file = NULL;

    while ((opt = getopt(argc, argv, "f:")) != -1) {
        switch (opt) {
        case 'f':
            input_file = optarg;
            break;
        default: /* '?' */
            printUsage(argv[0]);
        }
    }

    if (input_file == NULL) {
        printUsage(argv[0]);
    }

    return input_file;
}

void solution(char *fileName) {
    int floor = 0;
    int ch;

    FILE *file = fopen(fileName, "r");
    if (file == NULL) {
        perror("Error reopening file");
        exit(EXIT_FAILURE);
    }

    while ((ch = fgetc(file)) != EOF) {
        if (ch == '(') {
            floor++;
        } else if (ch == ')') {
            floor--;
        }
    }

    fclose(file);

    printf("Final floor: %d\n", floor);
}

// main is the entry point of the program. It processes command-line arguments and prints the input file name.
int main(int argc, char *argv[]) {
    printf("Advent of Code 2015 - Day 1\n");

    char* fileName = processArguments(argc, argv);

    printf("Input file: %s\n", fileName);

    solution(fileName);

    return 0;
}

File: 2015/go/day01/main.go
package main

import (
	"bufio"
	"flag"
	"io"
	"log"
	"os"
)

// solution reads the input file and calculates the final floor based on the parentheses.
func solution(inputFile string) {
	// Open the input file
	file, err := os.Open(inputFile)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	// Create a buffered reader to read the file
	reader := bufio.NewReader(file)

	var currentFloor int = 0

	// Read the file character by character
	for {
		char, _, err := reader.ReadRune()
		if err != nil {
			if err == io.EOF {
				break
			}
			log.Fatal(err)
		}

		if char == '(' {
			currentFloor++
		}
		if char == ')' {
			currentFloor--
		}
	}

	log.Printf("Current floor: %d", currentFloor)
}

// processArguments parses command-line flags and returns the input file path.
func processArguments() string {
	// if there are no command-line arguments, print usage and exit
	if len(os.Args) < 2 {
		log.Fatal("Usage: go run main.go -input-file <path_to_input_file>")
	}

	var inputFile string

	flag.StringVar(&inputFile, "input-file", "", "Path to the input file")
	flag.Parse()

	// if the input file flag is not provided, print usage and exit
	if inputFile == "" {
		log.Fatal("Please provide an input file using the -input-file or --input-file flag.")
	}

	return inputFile
}

// main is the entry point of the program. It parses command-line flags and calls the solution function.
func main() {
	inputFile := processArguments()

	log.Println("Advent of Code 2015 - Day 1")

	solution(inputFile)
}
